Raise MSRV to Rust 1.80

ErichDonGubler opened the pull request:
- [ ] Depends on Firefox updating to Rust 1.80 or later.
- ~~Depends on #6147.~~
- ~~Depends on #6180.~~
- Part of #6379.

ErichDonGubler committed:
chore: warn on `clippy::ref_as_ptr`

ErichDonGubler committed:
build: upgrade `cfg_aliases` 0.1.1 → 0.2.1

Consumes
[`cfg_aliases`#8](https://github.com/katharostech/cfg_aliases/pull/8) to
eliminate warnings from `rustc`.

ErichDonGubler committed:
chore: remove `std::mem::*` imports now unnecessary with Rust 1.80

`std::mem::{size,align}_of{,_val}` was added to `std::prelude` in Rust
1.80; see
[`rust`#123168](https://github.com/rust-lang/rust/pull/123168/).

ErichDonGubler committed:
chore(core): allow `wgpu_*` `cfg` keys not exposed in `features`

This resolves remaining outstanding cases that offend [`--check-cfg` in
Rust
1.80](https://blog.rust-lang.org/2024/07/25/Rust-1.80.0.html#checked-cfg-names-and-values)
in `wgpu-core`.

ErichDonGubler committed:
build: update Rust 1.78.0 → 1.80.1

ErichDonGubler commented:
Heeey, we're finally running into the same annoying errors [here](https://github.com/gfx-rs/wgpu/actions/runs/10351833426/job/28651170468?pr=6105#step:11:404) that I wasn't sure what to do with locally when I saw them:

```
error: trait `ContextId` is never used
  --> wgpu/src/context.rs:25:11
   |
25 | pub trait ContextId: Into<ObjectId> + From<ObjectId> + Debug + 'static {}
   |           ^^^^^^^^^
   |
   = note: `-D dead-code` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(dead_code)]`

error: methods `device_downlevel_properties`, `device_mark_lost`, `render_bundle_encoder_multi_draw_indirect`, `render_bundle_encoder_multi_draw_indexed_indirect`, `render_bundle_encoder_multi_draw_indirect_count`, and `render_bundle_encoder_multi_draw_indexed_indirect_count` are never used
   --> wgpu/src/context.rs:190:8
    |
34  | pub trait Context: Debug + WasmNotSendSync + Sized {
    |           ------- methods in this trait
...
190 |     fn device_downlevel_properties(
    |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
290 |     fn device_mark_lost(
    |        ^^^^^^^^^^^^^^^^
...
793 |     fn render_bundle_encoder_multi_draw_indirect(
    |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
802 |     fn render_bundle_encoder_multi_draw_indexed_indirect(
    |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
812 |     fn render_bundle_encoder_multi_draw_indirect_count(
    |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
825 |     fn render_bundle_encoder_multi_draw_indexed_indirect_count(
    |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: methods `device_downlevel_properties`, `device_mark_lost`, `render_bundle_encoder_multi_draw_indirect`, `render_bundle_encoder_multi_draw_indexed_indirect`, `render_bundle_encoder_multi_draw_indirect_count`, and `render_bundle_encoder_multi_draw_indexed_indirect_count` are never used
    --> wgpu/src/context.rs:1249:8
     |
1171 | pub(crate) trait DynContext: Debug + WasmNotSendSync {
     |                  ---------- methods in this trait
...
1249 |     fn device_downlevel_properties(
     |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
1349 |     fn device_mark_lost(&self, device: &ObjectId, device_data: &crate::Data, message: &str);
     |        ^^^^^^^^^^^^^^^^
...
1790 |     fn render_bundle_encoder_multi_draw_indirect(
     |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
1799 |     fn render_bundle_encoder_multi_draw_indexed_indirect(
     |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
1809 |     fn render_bundle_encoder_multi_draw_indirect_count(
     |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
1822 |     fn render_bundle_encoder_multi_draw_indexed_indirect_count(
     |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

    Checking naga-fuzz v0.0.0 (/Users/runner/work/wgpu/wgpu/naga/fuzz)
    Checking d3d12 v22.0.0 (/Users/runner/work/wgpu/wgpu/d3d12)
error: could not compile `wgpu` (lib) due to 3 previous errors
```

☝🏻 @cwfitzgerald, @Wumpf: Not sure how to best proceed here, but I haven't given it many brain cycles, either. Will dedicate more in the absence of feedback in the next day or so.

ErichDonGubler commented:
Leaving this as a draft until we actually have the option to upgrade to Rust 1.80.

We might be able to hoist the upgrade to `cfg_aliases` out if https://github.com/katharostech/cfg_aliases/pull/10/ merges and is released.

ErichDonGubler committed:
WIP: build: update Rust 1.76 → 1.78

ErichDonGubler commented:
Updated OP with a link to Firefox tracking of Rust 1.80 in `mozilla-central`.

kornelski commented:
`size_of` could be made backwards-compatible by importing it `use std::mem::size_of;`. Rust doesn't warn about shadowing the prelude (yet).

ErichDonGubler commented:
> `size_of` could be made backwards-compatible by importing it `use std::mem::size_of;`. Rust doesn't warn about shadowing the prelude (yet).

Relevant `clippy` lint proposal by an individual who we know in this community (👋🏻 hi @kpreid!): https://github.com/rust-lang/rust-clippy/issues/8439

kornelski commented on naga/src/back/msl/mod.rs:
```diff
@@ -668,6 +668,6 @@ pub fn write_string(
 #[test]
 fn test_error_size() {
-    use std::mem::size_of;
+    use size_of;
```
Looks like find-replace error

ErichDonGubler replied:
Fixed. 😅 Thanks!

ErichDonGubler committed:
chore(benches): add `tracy` to `lints.rust.unexpected_cfgs.check-cfg`

ErichDonGubler committed:
chore: add `web_sys_unstable_apis` to `lints.rust.unexpected_cfgs.check-cfg` in `wgpu_{hal,types}`

cwfitzgerald commented:
We can bump the repo msrv without the core msrv, so this doesn't need to block on ff

ErichDonGubler commented:
@cwfitzgerald: This PR contains strictly backwards-incompatible changes, including in `wgpu-core` and `wgpu-hal` here. That's why I've kept this a draft and split out backwards-compatible changes repeatedly, trying to make CI green. I suppose we could reasonably split out the `wgpu`- and `examples`-specific changes for a separately mergeable PR! I don't think I'll be allocating the bandwidth for that for at least a week, but I'd be happy to review somebody who might do that before I did.

ErichDonGubler commented:
Not sure why GL tests are failing. I suspect this PR is at fault, but I'm not sure yet.

cwfitzgerald reviewed (approved):
We seem to have UB in the GL backend that rust is now catching.

ErichDonGubler committed:
refactor: s/once_cell::Lazy/std::sync::LazyLock

Weaken our dependence on the `once_cell` crate by using functionality
from `std` instead that was upstreamed from `once_cell`, this time with
what's available in Rust 1.80+.

It's not yet possible to eliminate this dependency entirely, but do what
we can for now.

ErichDonGubler committed:
chore: satisfy `clippy::manual_c_str_literals`

This was previously an `allow`-by-default warning in Clippy's `pedantic`
group, but with Rust 1.83 it was promoted to a `warn`-by-default member
of its `complexity` group.

ErichDonGubler committed:
refactor: use more `c` string literals
